[terra-notification-dialog] MVB + dialog style updates (#1105)
cerner/terra-framework: packages/terra-notification-dialog/src/terra-dev-site/doc/notification-dialog/UpgradeGuide.2.doc.mdx

<Badge />

# Terra Notification Dialog Upgrade Guide
## Upgrading from version 3 to version 4

> The latest changes of the Notification Dialog reflect the Patient Safety Notification Classification standards. Due to this, we have changes the variants provided by the Notification  Dialog to reflect the pre-defined hazardous vs non-hazardous notification types. To align with these standard, the flexibility to change  header (now signalWord for custom) and dialog icon are no longer customizable for these pre-defined variant types. The `custom` variant will need to be used if the predefined signal word-icon combination for a variant does not meet your needs.

Reviewer: ```suggestion
The latest changes to the Notification Dialog reflect the further development of Patient Safety Notification Classification standards. We have changed the variants provided by the Notification Dialog to reflect the defined hazardous vs. non-hazardous notification types. To align with these standards, the ability to change the header and dialog icon is no longer available for the defined variant types. The `custom` variant will need to be used if the predefined signal word and icon combination for a variant does not meet your needs.
```

Author: fixed: https://github.com/cerner/terra-framework/pull/1105/commits/c45612d9b69bb2982f5ca4adbcad9abaf6a161f5#diff-242e2663ef4cc154665a5927ff0cd178R8

<Badge />

> # Terra Notification Dialog Upgrade Guide

Reviewer 2: ```suggestion
# Terra Notification Dialog Upgrade Guide

```

## Upgrading from version 3 to version 4

The latest changes to the Notification Dialog reflect the further development of Patient Safety Notification Classification standards. We have changed the variants provided by the Notification Dialog to reflect the defined hazardous vs. non-hazardous notification types. To align with these standards, the ability to change the header and dialog icon is no longer available for the defined variant types. The `custom` variant will need to be used if the predefined signal word and icon combination for a variant does not meet your needs.

> ### New Props

Reviewer 2: Add a line break after headings
```suggestion
### New Props

```

- Added `dialogTitle` prop
- Added `custom` prop which accepts an object of keys
  - `signalWord`
  - `iconClassName`

> ### Changed Props

Reviewer 2: ```suggestion
### Changed Props

```

> - `variant` prop is now required and the supported values have changes. The previous variants were `alert`, `warning`, `error`, `success`, `info`, and `custom`. The new variants are `hazard-high`, `hazard-medium`, `hazard-low`, `error`, and `custom`. The follows are the mappings for these new variants:

Reviewer 2: ```suggestion
- `variant` prop is now required and the supported values have changes. The previous variants were `alert`, `warning`, `error`, `success`, `info`, and `custom`. The new variants are `hazard-high`, `hazard-medium`, `hazard-low`, `error`, and `custom`. The mappings from the v3 variants to the v4 variants are as follows:
```

  - `alert` -> `hazard-high`
  - `warning` -> `hazard-medium`
  - `info` -> `hazard-low`
  - `error` and `custom` - these are the same.
  - `success` - this variant has been removed.

> - `isOpen` prop is no longer required

Reviewer 2: Do you want to state what to do instead of using is open? (basically only render the component if you want it open)

Author: This should have been removed when I removed the `isOpen` prop and added docs on it's removal. updated: https://github.com/cerner/terra-framework/pull/1105/commits/a29f0211211c048e519d4edc81b5e5d590e84dce#diff-242e2663ef4cc154665a5927ff0cd178L23

> ### Removed Props

Reviewer 2: ```suggestion
### Removed Props

```

- Removed `title` - use `dialogTitle` prop instead
- Removed `header` - create a custom notification dialog if a custom signal word is needed
- Removed `customIcon` - use `custom.iconClassName` prop instead
- Removed `primaryAction` - use `acceptAction` prop instead
- Removed `secondaryAction` - use `acceptAction` prop instead
- Removed `message` - use `startMessage` prop instead
- Removed `isOpen` - render NotificationDialog when it should be open

> ### Removed NotificationDialogVariants Export

Reviewer 2: ```suggestion
### Removed NotificationDialogVariants Export

```

The NotificationDialogVariants export was removed. Use the variant string instead.
```diff
- import NotificationDialog, { NotificationDialogVariants } from 'terra-notification-dialog';
+ import NotificationDialog from 'terra-notification-dialog';

const Dialog = () => {
  <NotificationDialog
-   variant={NotificationDialogVariants.ERROR}
+    variant="error"
..
```

### Changes to Creating Custom Dialogs

Use the prop `custom` to create a custom Notification Dialog and provided
- a keyword to the `signalWord` prop - previously referenced as header prop
- a icon as a background-image style via the `iconClassName` prop - previously provided as a React component using customIcon prop

Not providing these pieces will render a Custom Notification dialog with space allocated for the missing pieces.

**Note:** by choosing to create a custom Notification Dialog, you are responsible for providing an icon that is themed correctly in the terra-default-theme, orion-fusion-theme and in the clinical-lowlight-theme. See Terra's [Theme Strategy Guide](/guides/terra-application/theme-strategy) for more information.

> ##### Component.jsx

Reviewer 2: Don't jump heading levels :)
```suggestion
#### Component.jsx

```

```diff
  import NotificationDialog from 'terra-notification-dialog';
- import MyIcon from './MyIcon';
+ import ThemeContext from 'terra-theme-context';
+ import classnamesBind from 'classnames/bind';

+ import styles from './Component.module.scss';
+ cx = classnamesBind.bind(styles);

const Dialog = () => {
+ const theme = React.useContext(ThemeContext);

  return (
    <NotificationDialog
      variant="custom"
-     header="Horse Lovers"
-     customIcon={<MyIcon />}
+     custom={{
+       signalWord: 'Horse Lovers',
+       iconClassName: cx('my-icon', theme.className)
+     }}
      dialogTitle="Horses are the best"
      startMessage="Horses are super cool"
      acceptAction={{
        text: 'Super Interesting!',
        onClick: handleCloseModal,
      }}
    />
  );
 };
```

> ##### Component.module.scss

Reviewer 2: ```suggestion

#### Component.module.scss

```

```diff
+ @import '~terra-mixins/lib/Mixins';

:local {
+  .my-icon { 
+    background: var(--example-custom-icon-background, inline-svg("..."));
+  }

+  .orion-fusion-theme {
+    @include terra-inline-svg-var('--example-custom-icon-background', "...");
+  }

+  .clinical-lowlight-theme {
+    @include terra-inline-svg-var('--example-custom-icon-background', "...");
+  }
}
```

> ### Added Layout for Rendering Lists with content Prop

Reviewer 2: The markdown linter doesn't handle mdx files :(
```suggestion
### Added Layout for Rendering Lists with content Prop

```

The Notification Dialog provides has added the `ContentLayoutAsList` layout component to be used with the `content` prop. This layout renders a list of strings with styles that align with the Notification Dialog. Usage like:

``` diff
- import NotificationDialog from 'terra-notification-dialog';
+ import NotificationDialog, { ContentLayoutAsList } from 'terra-notification-dialog';

const Dialog = () => (
  <NotificationDialog
    variant="Info"
    dialogTitle="Did you know?"
    startMessage="That that cows:"
-   content={() => {
-     <ul>
-       <li key='info-1'>eat 40lbs of food a day</li>
-       <li key='info-1'>have no upper front teeth</li>
-     </ul>
-   }}
+   content={(
+     <ContentLayoutAsList
+        items={['eat 40lbs of food a day', 'have no upper front teeth ']}
+     />
+   )}
    endMessage="Interesting right?"
    acceptAction={{
      text: 'Super Interesting!',
      onClick: handleCloseModal,
    }}
  />
);
```

> ### Added Data Attributes for Testing

Reviewer 2: ```suggestion
### Added Data Attributes for Testing

```

The data attribute, `data-terra-notification-dialog`, was added to assist with validating the Notification Dialog was presented. This can be used like:

> ```diff

Reviewer 2: ```suggestion

```diff
```
